src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -659,47 +664,61 @@ public bool IsFrozen
         /// <returns></returns>
         internal bool IsAnyUpstreamFrozen()
         {
-            bool ret = false;
-            List<NodeModel> nodes = new List<NodeModel>();
-            return CheckIfAnyUpstreamNodeIsFrozen(this, nodes, ref ret);
+            return Upstream.Any(x => x.isFrozenExplicitly);
         }
 
-        private bool CheckIfAnyUpstreamNodeIsFrozen(NodeModel node, List<NodeModel> nodes, ref bool ret)
-        {             
-            var sets = node.InputNodes.Values;
-            var inpNodes = sets.Where(x => x != null).Select(z => z.Item2).Distinct();
-            foreach (var inode in inpNodes)
+        /// <summary>
+        /// Computes the upstream nodes for any given node.
+        /// </summary>
+        internal void ComputeUpstreamNodes()
```

Comment:
I think we need to either change the name or the comment or both, since this method also calculates upstream nodes for all downstream nodes of the node this method is called on.